In [307]:
import io
import json
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import requests
import seaborn as sb
from timeit import default_timer as timer
from tweepy import OAuthHandler
import xml.etree.ElementTree

# The Goal

Wrangle WeRateDogs Twitter data to create interesting and trustworthy analyses and visualizations. The Twitter archive is great, but it only contains very basic tweet information. Additional gathering, then assessing and cleaning is required for better analyses and visualizations.

# Table of contents
1. [Gathering Data](#Gathering)
    1. [WeRateDogs Twitter archive](#tweets)
    2. [Image Predictions](#predictions)
    3. [Tweets Info](#tweets_info)
2. [Assessing Data](#Assessing)
    1. [Tweets Info assessing](#info_assessing)
    2. [WeRateDogs Twitter archive assessing](#tweets_assessing)
    3. [Image Predictions assessing](#predictions_assessing)
    4. [Issues Summary](#Issues_summary)
3. [Cleaning Data](#cleaning_data)
4. [Storing The Data](#storing)
5. [insights](#insights)
    1. [insight 1](#info_assessing)
    2. [insight 2](#tweets_assessing)
    3. [insight 3](#predictions_assessing)

# Gathering Data <a name="Gathering"></a>

I will import 3 datasets into pandas dataframes:

1. WeRateDogs Twitter archive - by importing csv from the system
2. tweet image predictions - tcv file using requests
3. tetweet info - using tweepy and twitter's api


### WeRateDogs Twitter archive - from pc  <a name="tweets"></a>

In [41]:
# read simple csv file on pc
dogs_tweets = pd.read_csv('twitter-archive-enhanced.csv')

# validate reading
print(dogs_tweets.shape)
dogs_tweets.head(2)

(2356, 17)


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None  
1                  10    Tilly  None    None   None  None

### Tweets Image Predictions - from url Image <a name="predictions"></a>

In [42]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'


def request_to_df(url, sep='\t', header=0):
    '''
    desc:
    ---------
    request csv file and import it into df
    
    input:
    ----------
    * url - link to csv file
    * sep - a string, Delimiter to use, a pandas parameter. /t for tab 
    * header - Row number(s) to use as the column names. row 0 is default
    
    output:
    -----------
    Return a dataframe named df
    '''
    
    urlData = requests.get(url).content
    df = pd.read_csv(io.StringIO(urlData.decode('utf-8')), sep=sep, header=header)
    return df


predictions = request_to_df(url)

# validate reading
print(predictions.shape)
predictions.head(2)

(2075, 12)


tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True

### tweet info - from api to df <a name="tweets_info"></a>

In [ ]:
# Code from Udacity
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
# These are hidden to comply with Twitter's API terms and conditions

df_1 = dogs_tweets.copy()

consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True)

# NOTE TO STUDENT WITH MOBILE VERIFICATION ISSUES:
# df_1 is a DataFrame with the twitter_archive_enhanced.csv file. You may have to
# change line 17 to match the name of your DataFrame with twitter_archive_enhanced.csv
# NOTE TO REVIEWER: this student had mobile verification issues so the following
# Twitter API code was sent to this student from a Udacity instructor
# Tweet IDs for which to gather additional data via Twitter's API

tweet_ids = df_1.tweet_id.values
len(tweet_ids)

# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
count = 0
fails_dict = {}
start = timer()
# Save each tweet's returned JSON as a new line in a .txt file
with open('tweet_json.txt', 'w') as outfile:
    # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
    for tweet_id in tweet_ids:
        count += 1
        print(str(count) + ": " + str(tweet_id))
        try:
            tweet = api.get_status(tweet_id, tweet_mode='extended')
            print("Success")
            json.dump(tweet._json, outfile)
            outfile.write('\n')
        except tweepy.TweepError as e:
            print("Fail")
            fails_dict[tweet_id] = e
            pass
end = timer()
# print(end - start)
# print(fails_dict)

1: 892420643555336193
Fail
2: 892177421306343426
Fail
3: 891815181378084864
Fail
4: 891689557279858688
Fail
5: 891327558926688256
Fail
6: 891087950875897856
Fail
7: 890971913173991426
Fail
8: 890729181411237888
Fail
9: 890609185150312448
Fail
10: 890240255349198849
Fail
11: 890006608113172480
Fail
12: 889880896479866881
Fail
13: 889665388333682689
Fail
14: 889638837579907072
Fail
15: 889531135344209921
Fail
16: 889278841981685760
Fail
17: 888917238123831296
Fail
18: 888804989199671297
Fail
19: 888554962724278272
Fail
20: 888202515573088257
Fail
21: 888078434458587136
Fail
22: 887705289381826560
Fail
23: 887517139158093824
Fail
24: 887473957103951883
Fail
25: 887343217045368832
Fail
26: 887101392804085760
Fail
27: 886983233522544640
Fail
28: 886736880519319552
Fail
29: 886680336477933568
Fail
30: 886366144734445568
Fail
31: 886267009285017600
Fail
32: 886258384151887873
Fail
33: 886054160059072513
Fail
34: 885984800019947520
Fail
35: 885528943205470208
Fail
36: 885518971528720385
Fail
3

In [43]:
def json_to_df(txt_file):
    '''
    desc:
    ---------
    Convert json data from twiiter to into dataframe.
    The columns are hard coded for now
    
    input:
    ----------
    txt_file - a path to text file with json data from api
    
    output:
    -----------
    Return a dataframe named df
    '''
    
    tweet_info = []
    
    # read file into list
    with open(txt_file, "r") as f:
        for line in f:
            tweet = json.loads(line)
            tweet_info.append({
                "tweet_id": tweet["id"],
                "retweet_count": tweet["retweet_count"],
                "favorite_count": tweet["favorite_count"],"retweeted": tweet["retweeted"]                
            })
            
    # convert list into df
    df = pd.DataFrame(tweet_info, 
                      columns = ["tweet_id", "retweet_count", "favorite_count"])
    return df

In [44]:
tweet_info = json_to_df("tweet-json.txt")

# validate reading
print(tweet_info.shape)
tweet_info.head(5)

(2354, 3)


tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

# Assessing Data <a name="Assessing"></a>

I will assess the data from the 3 data frames visually and programmatically for quality and tidiness issues.
I will docuement each issue along the way and in the end of this part.

The data frame will be tested in following order:
1. tweets info
2. dogs tweets
3. predictions

## Tweets information <a name="info_assessing"></a>

In [45]:
display(tweet_info.head())
display(tweet_info.tail())

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467
1  892177421306343426           6514           33819
2  891815181378084864           4328           25461
3  891689557279858688           8964           42908
4  891327558926688256           9774           41048

tweet_id  retweet_count  favorite_count
2349  666049248165822465             41             111
2350  666044226329800704            147             311
2351  666033412701032449             47             128
2352  666029285002620928             48             132
2353  666020888022790149            532            2535

In [46]:
tweet_info.info()
tweet_info.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   tweet_id        2354 non-null   int64
 1   retweet_count   2354 non-null   int64
 2   favorite_count  2354 non-null   int64
dtypes: int64(3)
memory usage: 55.3 KB


tweet_id  retweet_count  favorite_count
count  2.354000e+03    2354.000000     2354.000000
mean   7.426978e+17    3164.797366     8080.968564
std    6.852812e+16    5284.770364    11814.771334
min    6.660209e+17       0.000000        0.000000
25%    6.783975e+17     624.500000     1415.000000
50%    7.194596e+17    1473.500000     3603.500000
75%    7.993058e+17    3652.000000    10122.250000
max    8.924206e+17   79515.000000   132810.000000

In [47]:
tweet_info[tweet_info["tweet_id"].duplicated()]

Empty DataFrame
Columns: [tweet_id, retweet_count, favorite_count]
Index: []

1. There are no any duplicated values, weird max / min values 
2. There isn't any null at all. 
3. The tweet_id column need to be convert to string.

## dogs tweets <a name="tweets_assessing"></a>

In [48]:
dogs_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

1. tweet_id need to be converted to str, because it is a key.
2. timestamp can be converted into date data type.
3. most of the fields retweeted_status_id, retweeted_status_user_id, retweeted_status_timestamp, in_reply_to_status_id and in_reply_to_user_id have missing values. They maybe can be removed. If I will keep them, they need to str due to them id field, and the timestamp need to convert into timestamp.

In [49]:
# print all rows for trying to assest visually

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(dogs_tweets)

                tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                 

1. The source - can probaly remove the link and the html tags, just need to make sure there is different sources.
2. Some dogs don't have names or have weird name, like 'a' or 'the'. I will remove the names a, none and the.
3. not all dogs have type.

The first issue is a tidy issue - it seems the columns doggo, floofer, pupper and puppo can be one field. each include two possibiltes, the field name or None.

In [50]:
# Does the dog type field are binary? 

print(dogs_tweets.puppo.unique())
print(dogs_tweets.pupper.unique())
print(dogs_tweets.doggo.unique())
print(dogs_tweets.floofer.unique())

['None' 'puppo']
['None' 'pupper']
['None' 'doggo']
['None' 'floofer']


In [51]:
# Are There more then 1 type of source?

dogs_tweets.source.unique()

array(['<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>',
       '<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>',
       '<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>',
       '<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>'],
      dtype=object)

In [52]:
# check for weird dog's names

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print(dogs_tweets.name.value_counts())

None              745
a                  55
Charlie            12
Lucy               11
Oliver             11
Cooper             11
Penny              10
Tucker             10
Lola               10
Winston             9
Bo                  9
the                 8
Sadie               8
Daisy               7
Toby                7
Buddy               7
Bailey              7
an                  7
Leo                 6
Milo                6
Jack                6
Rusty               6
Scout               6
Jax                 6
Bella               6
Oscar               6
Dave                6
Stanley             6
Koda                6
Phil                5
Gus                 5
Sunny               5
Alfie               5
Chester             5
Louis               5
Bentley             5
Oakley              5
very                5
George              5
Sammy               5
Finn                5
Larry               5
Maximus             4
Walter              4
Boomer              4
Bruce     

In [21]:
# How many dogs with 1 letter name?

dogs_tweets[dogs_tweets['name'].apply(len)==1][['tweet_id','name']].groupby('name').count()

tweet_id
name          
O            1
a           55

In [22]:
# How many dogs with 2 letter name?

dogs_tweets[dogs_tweets['name'].apply(len)==2][['tweet_id','name']].groupby('name').count()

tweet_id
name          
Al           1
Bo           9
Ed           1
JD           1
Jo           1
Mo           1
an           7
by           1
my           1

In [24]:
# How many dogs with 3 letter name?

dogs_tweets[dogs_tweets['name'].apply(len)==3][['tweet_id','name']].groupby('name').count()

tweet_id
name          
Ace          1
Aja          1
Alf          1
Amy          1
Ash          2
Ava          2
Ben          1
Blu          1
Bob          2
Cal          1
Dex          1
Doc          2
Dot          1
Dug          1
Edd          1
Eli          2
Eve          2
Evy          1
Gin          1
Gus          5
Ike          1
Ito          1
Jax          6
Jay          1
Jeb          1
Jed          1
Jim          1
Ken          2
Leo          6
Lou          2
Mac          1
Max          3
Mia          3
Moe          2
Mya          1
Obi          1
Ole          1
Pip          1
Rey          1
Ron          1
Sam          2
Sid          1
Sky          1
Stu          1
Taz          1
Ted          3
Tom          1
Tug          1
Tyr          2
Zoe          1
all          1
his          1
mad          2
not          2
old          1
one          4
the          8

In [33]:
# find all non capitlize / all letters are lovwer case

with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(dogs_tweets[dogs_tweets.name.str.islower()==True][['tweet_id', 'name']])

tweet_id          name
22    887517139158093824          such
56    881536004380872706             a
118   869988702071779329         quite
169   859196978902773760         quite
193   855459453768019968         quite
335   832645525019123713           not
369   828650029636317184           one
542   806219024703037440    incredibly
649   792913359805018113             a
682   788552643979468800           mad
759   778396591732486144            an
773   776249906839351296          very
801   772581559778025472             a
819   770655142660169732          very
822   770093767776997377          just
852   765395769549590528            my
924   755206590534418437           one
988   748977405889503236           not
992   748692773788876800           his
993   748575535303884801           one
1002  747885874273214464             a
1004  747816857231626240             a
1017  746872823977771008             a
1025  746369468511756288            an
1031  745422732645535745          very
1040  744223424764059648      actually
1049  743222593470234624             a
1063  741067306818797568          just
1071  740214038584557568       getting
1095  736392552031657984           mad
1097  736225175608430592          very
1120  731156023742988288          this
1121  730924654643314689  unacceptable
1138  728035342121635841           all
1193  717537687239008257             a
1206  715758151270801409           old
1207  715733265223708672             a
1259  710272297844797440   infuriating
1340  704859558691414016             a
1351  704054845121142784             a
1361  703079050210877440             a
1362  703041949650034688            an
1368  702539513671897089             a
1382  700864154249383937             a
1385  700747788515020802          very
1435  697259378236399616       getting
1457  695095422348574720          just
1499  692187005137076224             a
1527  690360449368465409           the
1603  685943807276412928           the
1693  681297372102656000      actually
1724  680085611152338944            by
1737  679530280114372609             a
1747  679111216690831360    officially
1785  677644091929329666             a
1797  677269281705472000           the
1815  676613908052996102           the
1853  675706639471788032             a
1854  675534494439489536             a
1877  675109292475830276             a
1878  675047298674663426             a
1916  674307341513269249          life
1923  674082852460433408             a
1936  673956914389192708           one
1941  673715861853720576             a
1955  673636718965334016             a
1994  672604026190569472             a
2001  672482722825261057         light
2019  672125275208069120          just
2030  671789708968640512         space
2034  671743150407421952             a
2037  671561002136281088           the
2066  671147085991960577             a
2116  670427002554466305             a
2125  670361874861563904             a
2128  670303360680108032             a
2146  669923323644657664             a
2153  669661792646373376             a
2161  669564461267722241             a
2191  668955713004314625             a
2198  668815180734689280             a
2204  668636665813057536            an
2211  668614819948453888             a
2212  668587383441514497           the
2218  668507509523615744             a
2222  668466899341221888             a
2235  668171859951755264             a
2249  667861340749471744             a
2255  667773195014021121             a
2264  667538891197542400             a
2273  667470559035432960             a
2287  667177989038297088             a
2304  666983947667116034             a
2311  666781792255496192             a
2314  666701168228331520             a
2326  666411507551481857         quite
2327  666407126856765440             a
2333  666337882303524864            an
2334  666293911632134144             a
2335  666287406224695296            an
2345  666063827256086533           the
2346  666058600524156928           the

In [35]:
# How many dogs do not have real name?

dogs_tweets[dogs_tweets.name.str.islower()==True][['name']].count()

name    109
dtype: int64

Werid / none names all be remove so we all be able to analyze dogs names better. It seems the easist way is to remove all non capitlize names based if all is lower case

names example: 
not, all, my, by, old, an, a, O, None, the, very, this, his, such, quite, incredibly, mad, getting, actually, infuriatingm officially  

In [13]:
# understaning the the ratings columns 
dogs_tweets[['rating_numerator', 'rating_denominator']].describe()

rating_numerator  rating_denominator
count       2356.000000         2356.000000
mean          13.126486           10.455433
std           45.876648            6.745237
min            0.000000            0.000000
25%           10.000000           10.000000
50%           11.000000           10.000000
75%           12.000000           10.000000
max         1776.000000          170.000000

1. The avg numerator is bigger then denominator which is correct.
2. There are dogs with 0 numerator and/or 0 denominator, which need checking.
3. max denominator of 170 - need testing

In [39]:
# find rating rating numerator == 0
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets.query("rating_numerator == 0")[['tweet_id','text', 'expanded_urls', 'rating_numerator', 'rating_denominator']])

tweet_id  \
315   835152434251116546   
1016  746906459439529985   

                                                                                                                         text  \
315         When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag   
1016  PUPDATE: can't see any. Even if I could, I couldn't reach them to pet. 0/10 much disappointment https://t.co/c7WXaB2nqX   

                                                                                                                                                                                        expanded_urls  \
315   https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1,https://twitter.com/dog_rates/status/835152434251116546/photo/1   
1016                                                                                                                                  https://twitter.com/dog_rates/status/746906459439529985/photo/1   

      rating_numerator  rating_denominator  
315                  0                  10  
1016                 0                  10

In [55]:
# find rating denominator == 0
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets.query("rating_denominator == 0")[['tweet_id', 'expanded_urls', 'text', 'rating_numerator', 'rating_denominator']])

tweet_id expanded_urls  \
313  835246439529840640           NaN   

                                                                                                   text  \
313  @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   

     rating_numerator  rating_denominator  
313               960                   0

1. Ids 835152434251116546 and 746906459439529985 can be removed due to them not dog tweet.
2. Id 835246439529840640 seem more as a joke then real rating - will be removed

In [36]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets[dogs_tweets.rating_denominator > 10][['text', 'expanded_urls', 'rating_numerator', 'rating_denominator']])

text  \
342                                                                                                             @docmisterio account started on 11/15/15   
433                                                  The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd   
784         RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…   
902                                                                       Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE   
1068        After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   
1120                           Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv   
1165                                                                                    Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a   
1202                               This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq   
1228                                                       Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1   
1254                                        Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12   
1274  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
1351                                                                  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa   
1433                                                        Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ   
1598            Yes I do realize a rating of 4/20 would've been fitting. However, it would be unjust to give these cooperative pups that low of a rating   
1634                      Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3   
1635        Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55   
1662         This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5   
1663         I'm aware that I could've said 20/16, but here at WeRateDogs we are very professional. An inconsistent rating scale is simply irresponsible   
1779                                                                              IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq   
1843                                         Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw   

                                                                                                                                                                                                expanded_urls  \
342                                                                                                                                                                                                       NaN   
433           https://twitter.com/dog_rates/status/820690176645140481/photo/1,https://twitter.com/dog_rates/status/820690176645140481/photo/1,https://twitter.com/dog_rates/status/820690176645140481/photo/1   
784   https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,http...   
902                                

In [54]:
# 342 and 1663 can be removed - not dogs?
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets.iloc[[342, 1663]][['text', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name']])

text  \
342                                                                                                      @docmisterio account started on 11/15/15   
1663  I'm aware that I could've said 20/16, but here at WeRateDogs we are very professional. An inconsistent rating scale is simply irresponsible   

     expanded_urls  rating_numerator  rating_denominator  name  
342            NaN                11                  15  None  
1663           NaN                20                  16  None

In [37]:
# 1068 and 784 the same dog? maybe all 'RT @dog_rates' need to be removed
 with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets.iloc[[784, 1068]][['text', 'expanded_urls', 'rating_numerator', 'rating_denominator', 'name']])

text  \
784   RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…   
1068  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   

                                                                                                                                                                                                expanded_urls  \
784   https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,http...   
1068  https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,https://twitter.com/dog_rates/status/740373189193256964/photo/1,http...   

      rating_numerator  rating_denominator  name  
784                  9                  11  None  
1068                 9                  11  None

In [46]:
# check for numerator smaller then denominator

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets[dogs_tweets.rating_denominator > dogs_tweets.rating_numerator][
        ['text', 'rating_numerator', 'rating_denominator']])

text  \
45                              This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948   
229   This is Jerry. He's doing a distinguished tongue slip. Slightly patronizing tbh. You think you're better than us, Jerry? 6/10 hold me back https://t.co/DkOBbwulw1   
315                                                    When you're so blinded by your systematic plagiarism that you forget what day it is. 0/10 https://t.co/YbEJPkg4Ag   
342                                                                                                                             @docmisterio account started on 11/15/15   
387                                                                                                I was going to do 007/10, but the joke wasn't worth the &lt;10 rating   
462                         RT @dog_rates: Meet Herschel. He's slightly bigger than ur average pupper. Looks lonely. Could probably ride 7/10 would totally pet https:/…   
485                     RT @dog_rates: Meet Beau &amp; Wilbur. Wilbur stole Beau's bed from him. Wilbur now has so much room for activities. 9/10 for both pups https:/…   
599                         RT @dog_rates: Here we see a rare pouched pupper. Ample storage space. Looks alert. Jumps at random. Kicked open that door. 8/10 https://t.…   
605                         RT @dog_rates: Not familiar with this breed. No tail (weird). Only 2 legs. Doesn't bark. Surprisingly quick. Shits eggs. 1/10 https://t.co/…   
730                                           Who keeps sending in pictures without dogs in them? This needs to stop. 5/10 for the mediocre road https://t.co/ELqelxWMrC   
745                         RT @dog_rates: This is Hank. He's mischievous af. Doesn't even know what he was trying to do here. 8/10 quit the shit Hank damn https://t.c…   
764                         RT @dog_rates: Meet Gerald. He's a fairly exotic doggo. Floofy af. Inadequate knees tho. Self conscious about large forehead. 8/10 https://…   
765                         This is Wesley. He's clearly trespassing. Seems rather h*ckin violent too. Weaponized forehead. 3/10 wouldn't let in https://t.co/pL7wbMRW7M   
784                         RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…   
814                         Another pic without a dog in it? What am I supposed to do? Rate the carpet? Fine I will. 7/10 looks adequately comfy https://t.co/OJZQ6I4gGd   
835                          Meet Gerald. He's a fairly exotic doggo. Floofy af. Inadequate knees tho. Self conscious about large forehead. 8/10 https://t.co/WmczvjCWJq   
845                                This is Shadoe. Her tongue flies out of her mouth at random. Can't have a serious conversation with her. 9/10 https://t.co/Tytt15VquG   
859                                       This is Clark. He collects teddy bears. It's absolutely h*ckin horrifying. 8/10 please stop this Clark https://t.co/EDMcwt86fU   
860                         RT @dog_rates: Meet Eve. She's a raging alcoholic 8/10 (would b 11/10 but pupper alcoholism is a tragic issue that I can't condone) https:/…   
883                         This is Fido. He can tell the weather. Not good at fetch tho. Never comes when called. 4/10 would probably still pet https://t.co/4gOv2Q3iKP   
896                         Meet Toby. He has a drinking problem. Inflatable marijuana plant in the back is also not a good look. 7/10 cmon Toby https://t.co/Cim4DSj6Oi   
906                         Teagan reads entire books in store so they're free. Loved 50 Shades of Grey (how dare I make that joke so late) 9/10 https://t.co/l46jwv5WYv   
912                         Here's another picture without a dog in it. Idk why you guys keep sending these. 4/10 just because that's a neat rug https://t.co/mOmnL19Wsl   
915                           This i

Because we can see ratings who are float let's check how much it happend. Anyway it will need fixing.

In [53]:
# check for float ratings

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets[dogs_tweets['text'].str.contains(r"(\d+\.\d*\/\d+)")][
        ['text', 'rating_numerator', 'rating_denominator']])

C:\Users\user\anaconda3\envs\Udacity\lib\site-packages\pandas\core\strings.py:2001: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


text  \
45                             This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948   
340                        RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…   
695                                    This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
1689                                                         I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace   
1712                                                         Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD   

      rating_numerator  rating_denominator  
45                   5                  10  
340                 75                  10  
695                 75                  10  
763                 27                  10  
1689                 5                  10  
1712                26                  10

## Dogs Type Predections <a name="predictions_assessing"></a>

In [56]:
predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [59]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(predictions)

tweet_id  \
0     666020888022790149   
1     666029285002620928   
2     666033412701032449   
3     666044226329800704   
4     666049248165822465   
5     666050758794694657   
6     666051853826850816   
7     666055525042405380   
8     666057090499244032   
9     666058600524156928   
10    666063827256086533   
11    666071193221509120   
12    666073100786774016   
13    666082916733198337   
14    666094000022159362   
15    666099513787052032   
16    666102155909144576   
17    666104133288665088   
18    666268910803644416   
19    666273097616637952   
20    666287406224695296   
21    666293911632134144   
22    666337882303524864   
23    666345417576210432   
24    666353288456101888   
25    666362758909284353   
26    666373753744588802   
27    666396247373291520   
28    666407126856765440   
29    666411507551481857   
30    666418789513326592   
31    666421158376562688   
32    666428276349472768   
33    666430724426358785   
34    666435652385423360   
35    666437273139982337   
36    666447344410484738   
37    666454714377183233   
38    666644823164719104   
39    666649482315059201   
40    666691418707132416   
41    666701168228331520   
42    666739327293083650   
43    666776908487630848   
44    666781792255496192   
45    666786068205871104   
46    666804364988780544   
47    666817836334096384   
48    666826780179869698   
49    666835007768551424   
50    666837028449972224   
51    666983947667116034   
52    666996132027977728   
53    667012601033924608   
54    667044094246576128   
55    667062181243039745   
56    667065535570550784   
57    667073648344346624   
58    667090893657276420   
59    667119796878725120   
60    667138269671505920   
61    667152164079423490   
62    667160273090932737   
63    667165590075940865   
64    667171260800061440   
65    667174963120574464   
66    667176164155375616   
67    667177989038297088   
68    667182792070062081   
69    667188689915760640   
70    667192066997374976   
71    667200525029539841   
72    667211855547486208   
73    667369227918143488   
74    667393430834667520   
75    667405339315146752   
76    667435689202614272   
77    667437278097252352   
78    667443425659232256   
79    667453023279554560   
80    667455448082227200   
81    667470559035432960   
82    667491009379606528   
83    667495797102141441   
84    667502640335572993   
85    667509364010450944   
86    667517642048163840   
87    667524857454854144   
88    667530908589760512   
89    667534815156183040   
90    667538891197542400   
91    667544320556335104   
92    667546741521195010   
93    667549055577362432   
94    667550882905632768   
95    667550904950915073   
96    667724302356258817   
97    667728196545200128   
98    667766675769573376   
99    667773195014021121   
100   667782464991965184   
101   667793409583771648   
102   667801013445750784   
103   667806454573760512   
104   667832474953625600   
105   667861340749471744   
106   667866724293877760   
107   667873844930215936   
108   667878741721415682   
109   667885044254572545   
110   667886921285246976   
111   667902449697558528   
112   667911425562669056   
113   667915453470232577   
114   667924896115245057   
115   667937095915278337   
116   668113020489474048   
117   668142349051129856   
118   668154635664932864   
119   668171859951755264   
120   668190681446379520   
121   668204964695683073   
122   668221241640230912   
123   668226093875376128   
124   668237644992782336   
125   668248472370458624   
126   668256321989451776   
127   668268907921326080   
128   668274247790391296   
129   668286279830867968   
130   668291999406125056   
131   668297328638447616   
132   668466899341221888   
133   668480044826800133   
134   668484198282485761   
135   668496999348633600   
136   668507509523615744   
137   668528771708952576   
138   668537837512433665   
139   668542336805281792   
140   668544745690562560   
141   668567822092664832   
142   668614

In [65]:
# check duplicates

display(predictions['tweet_id'].duplicated().sum())
display(predictions['jpg_url'].duplicated().sum())

0

66

In [70]:
# Lets look at all the duplicates

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(predictions[predictions['jpg_url'].duplicated()])

tweet_id  \
1297  752309394570878976   
1315  754874841593970688   
1333  757729163776290825   
1345  759159934323924993   
1349  759566828574212096   
1364  761371037149827077   
1368  761750502866649088   
1387  766078092750233600   
1407  770093767776997377   
1417  771171053431250945   
1427  772615324260794368   
1446  775898661951791106   
1453  776819012571455488   
1456  777641927919427584   
1463  778396591732486144   
1476  780496263422808064   
1487  782021823840026624   
1495  783347506784731136   
1510  786036967502913536   
1522  788070120937619456   
1538  790723298204217344   
1541  791026214425268224   
1564  793614319594401792   
1569  794355576146903043   
1571  794983741416415232   
1579  796177847564038144   
1588  798340744599797760   
1589  798628517273620480   
1590  798644042770751489   
1591  798665375516884993   
1592  798673117451325440   
1593  798694562394996736   
1594  798697898615730177   
1601  799774291445383169   
1605  800443802682937345   
1615  802247111496568832   
1619  802624713319034886   
1624  803692223237865472   
1627  804413760345620481   
1634  805958939288408065   
1636  806242860592926720   
1640  807059379405148160   
1645  808134635716833280   
1652  809808892968534016   
1683  813944609378369540   
1693  816014286006976512   
1699  816829038950027264   
1703  817181837579653120   
1712  818588835076603904   
1717  819015331746349057   
1718  819015337530290176   
1727  820446719150292993   
1736  821813639212650496   
1742  822647212903690241   
1746  823269594223824897   
1755  824796380199809024   
1789  829878982036299777   
1803  832040443403784192   
1804  832215726631055365   
1858  841833993020538882   
1864  842892208864923648   
1903  851953902622658560   
1944  861769973181624320   
1992  873697596434513921   
2041  885311592912609280   
2055  888202515573088257   

                                                                                      jpg_url  \
1297  https://pbs.twimg.com/ext_tw_video_thumb/675354114423808004/pu/img/qL1R_nGLqa6lmkOx.jpg   
1315                                          https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1333                                          https://pbs.twimg.com/media/CWyD2HGUYAQ1Xa7.jpg   
1345                                          https://pbs.twimg.com/media/CU1zsMSUAAAS0qW.jpg   
1349                                          https://pbs.twimg.com/media/CkNjahBXAAQ2kWo.jpg   
1364                              https://pbs.twimg.com/tweet_video_thumb/CeBym7oXEAEWbEg.jpg   
1368                                          https://pbs.twimg.com/media/CYLDikFWEAAIy1y.jpg   
1387                                          https://pbs.twimg.com/media/ChK1tdBWwAQ1flD.jpg   
1407                                          https://pbs.twimg.com/media/CkjMx99UoAM2B1a.jpg   
1417                                          https://pbs.twimg.com/media/CVgdFjNWEAAxmbq.jpg   
1427                                          https://pbs.twimg.com/media/Cp6db4-XYAAMmqL.jpg   
1446                                          https://pbs.twimg.com/media/CiyHLocU4AI2pJu.jpg   
1453                                          https://pbs.twimg.com/media/CW88XN4WsAAlo8r.jpg   
1456                                          https://pbs.twimg.com/media/CmoPdmHW8AAi8BI.jpg   
1463                                          https://pbs.twimg.com/media/CcG07BYW0AErrC9.jpg   
1476                                          https://pbs.twimg.com/media/Ck2d7tJWUAEPTL3.jpg   
1487                                          https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
1495                                          https://pbs.twimg.com/media/CVuQ2LeUsAAIe3s.jpg   
1510                                          https://pbs.twimg.com/media/CtKHLuCWYAA2TTs.jpg   
1522                                          https://pbs.twimg.com/media/Co-hmcYXYAASkiG.jpg   
1538                                          https://pbs.twimg.com/media/CvaYgDOWgAEfjls.jpg   
1541 

In [71]:
# One duplicate for example

url='https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg'
with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(predictions[predictions['jpg_url']==url])

tweet_id                                          jpg_url  \
591   679158373988876288  https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   
1315  754874841593970688  https://pbs.twimg.com/media/CWza7kpWcAAdYLc.jpg   

      img_num   p1   p1_conf  p1_dog            p2  p2_conf  p2_dog  \
591         1  pug  0.272205    True  bull_mastiff  0.25153    True   
1315        1  pug  0.272205    True  bull_mastiff  0.25153    True   

              p3   p3_conf  p3_dog  
591   bath_towel  0.116806   False  
1315  bath_towel  0.116806   False

In [66]:
# check for wierd numbers, like for img_num

predictions.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

### Issues summary: 
<a name="Issues_summary"></a>

* Tidy issues
    1. **for all datasets** convert id column to str 
    2. **Dog tweets:** timestamp to date
    3. **Dog tweets:** dog types need to change to one field (using pivot table)
* Quality issues
    1. **Dog tweets:** source field remove html tags
    2. **Dog tweets:** Remove weird  and not names names (using all lower)
    3. **Dog tweets:** dog type None need to be convert to NaN
    4. **Dog tweets:** Remove non dogs ratings = > 835152434251116546, 746906459439529985, 835246439529840640
    5. **Dog tweets:** Fix incorrect ratings
    6. **Dog tweets:** remove retweets as they can be the same dog
    7. **Dog tweets:** Extended url name need to be change to tweet_url, so it will be more clear
    8. **Predictions:** Remove duplicate urls
    9. **Predictions:** change for better name, like prediction_jpg_url

# Cleaning Data 
<a name="cleaning_data"></a>

At this part I will clean the datasets, based on the resulst I got in the former part. The dataframes will first copied and then will be cleaned.
The format of this part is: 

**Define**

What the issue?

**Code**

The code itself

**Test**

Testing the code did his job

In [217]:
# Copy the datasets

tweet_info_clean = tweet_info.copy()
dogs_tweets_clean = dogs_tweets.copy()
predictions_clean = predictions.copy()

# test the copy work

display(tweet_info_clean.head(1))
display(dogs_tweets_clean.head(1))
display(predictions_clean.head(1))

tweet_id  retweet_count  favorite_count
0  892420643555336193           8853           39467

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name doggo floofer pupper puppo  
0                  10  Phineas  None    None   None  None

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   

   img_num                      p1   p1_conf  p1_dog      p2   p2_conf  \
0        1  Welsh_springer_spaniel  0.465074    True  collie  0.156665   

   p2_dog                 p3   p3_conf  p3_dog  
0    True  Shetland_sheepdog  0.061428    True

## Tidy Issues

**Define**

For all dataframes id field is int, it need to be converted to string, due to it not use for calculation.

**Code**

In [218]:
tweet_info_clean['tweet_id'] = tweet_info_clean['tweet_id'].astype(str)
dogs_tweets_clean['tweet_id'] = dogs_tweets_clean['tweet_id'].astype(str)
predictions_clean['tweet_id'] = predictions_clean['tweet_id'].astype(str)

**Test**

In [219]:
display(tweet_info_clean.info())
display(dogs_tweets_clean.info())
display(predictions_clean.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2354 entries, 0 to 2353
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   tweet_id        2354 non-null   object
 1   retweet_count   2354 non-null   int64 
 2   favorite_count  2354 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 55.3+ KB


None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   object 
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

None

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   object 
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 152.1+ KB


None

**Define**

For dogs tweets dataframe date fields is object, it need to be converted to date, so it will be more useful.

**Code**

In [220]:
# format 2017-08-01 16:23:56 +0000
dogs_tweets_clean['timestamp'] =  pd.to_datetime(dogs_tweets_clean['timestamp'])

**Test**

In [221]:
dogs_tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2356 non-null   object             
 1   in_reply_to_status_id       78 non-null     float64            
 2   in_reply_to_user_id         78 non-null     float64            
 3   timestamp                   2356 non-null   datetime64[ns, UTC]
 4   source                      2356 non-null   object             
 5   text                        2356 non-null   object             
 6   retweeted_status_id         181 non-null    float64            
 7   retweeted_status_user_id    181 non-null    float64            
 8   retweeted_status_timestamp  181 non-null    object             
 9   expanded_urls               2297 non-null   object             
 10  rating_numerator            2356 non-null   int64           

**Define**

In the dog tweets dataset there are 4 fields to dog type. It will be better with one column, based on the tidy rules.
First None will be need to be replaced for NaN

**Code**

In [222]:
dogs_tweets_clean.replace('None', np.NaN, inplace=True)

# testing

print(dogs_tweets_clean.puppo.unique())
print(dogs_tweets_clean.pupper.unique())
print(dogs_tweets_clean.doggo.unique())
print(dogs_tweets_clean.floofer.unique())

[nan 'puppo']
[nan 'pupper']
[nan 'doggo']
[nan 'floofer']


In [223]:
# melt all dogs type to one field. 
# WARNING! to run this cell again, one need to run the copy function again!

dog_columns_values = dogs_tweets_clean.columns[:-4].tolist()
dog_columns_vars = dogs_tweets_clean.columns[-4:].tolist()

dogs_tweets_clean = pd.melt(dogs_tweets_clean, id_vars=dog_columns_values, value_vars=dog_columns_vars, var_name='dog_type')
dogs_tweets_clean.drop(columns=['dog_type'], inplace=True)

In [224]:
dogs_tweets_clean = dogs_tweets_clean.drop_duplicates()

**Test**

In [225]:
print(dogs_tweets_clean.shape)
print(dogs_tweets_clean.value.unique())
dogs_tweets_clean.head(1)

(2750, 14)
[nan 'doggo' 'floofer' 'pupper' 'puppo']


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                  timestamp  \
0 2017-08-01 16:23:56+00:00   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   

   rating_denominator     name value  
0                  10  Phineas   NaN

Probably the are dogs with more then one type

In [226]:
duplicate_dogs_types = dogs_tweets_clean[dogs_tweets_clean.tweet_id.duplicated()][['tweet_id', 'value']]
duplicate_dogs_types

tweet_id    value
2365  890240255349198849      NaN
2399  884162670584377345      NaN
2402  883360690899218434  floofer
2455  872967104147763200      NaN
2464  871515927908634625      NaN
...                  ...      ...
8015  752519690950500352    puppo
8029  751132876104687617    puppo
8103  744995568523612160    puppo
8116  743253157753532416    puppo
8151  738537504001953792    puppo

[394 rows x 2 columns]

In [227]:
# lets test one case
dogs_tweets_clean[dogs_tweets_clean.tweet_id == '890240255349198849']

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
9     890240255349198849                    NaN                  NaN   
2365  890240255349198849                    NaN                  NaN   

                     timestamp  \
9    2017-07-26 15:59:51+00:00   
2365 2017-07-26 15:59:51+00:00   

                                                 source  \
9     <a href="http://twitter.com/download/iphone" r...   
2365  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
9     This is Cassie. She is a college pup. Studying...                  NaN   
2365  This is Cassie. She is a college pup. Studying...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
9                          NaN                        NaN   
2365                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
9     https://twitter.com/dog_rates/status/890240255...                14   
2365  https://twitter.com/dog_rates/status/890240255...                14   

      rating_denominator    name  value  
9                     10  Cassie  doggo  
2365                  10  Cassie    NaN

In [228]:
# How many dogs of each type?

duplicate_dogs_types.value.value_counts()

pupper     257
puppo       30
floofer     10
Name: value, dtype: int64

In [229]:
# There are duplicated when droping only NaN?

duplicate_dogs_types.dropna().duplicated().sum()

0

In [230]:
# drop all NaN duplicated dogs

drop_duplicates_arr = duplicate_dogs_types[duplicate_dogs_types['value'].isna()].index
dogs_tweets_clean = dogs_tweets_clean.drop(drop_duplicates_arr)
dogs_tweets_clean.duplicated().sum()

0

In [231]:
dogs_tweets_clean.shape

(2653, 14)

In [254]:
dogs_tweets_clean.rename(columns={'value':'dog_type'}, inplace=True)

**Define**

In the dog tweets dataset the source field include html tags, there are not relevent for our analsys

**Code**

In [248]:
def remove_tags(text):
    '''
    desc:
    ---------
    Clean html tags from string
    
    input:
    ----------
    * text - a string (with html tags)
    
    output:
    -----------
    Return a string without html tags
    '''
    
    return ''.join(xml.etree.ElementTree.fromstring(text).itertext())

In [247]:
dogs_tweets_clean['source'] = dogs_tweets_clean.source.apply(remove_tags)

**Test**

In [249]:
dogs_tweets_clean.source.unique()

array(['Twitter for iPhone', 'Twitter Web Client', 'Vine - Make a Scene',
       'TweetDeck'], dtype=object)

**Define**

In the dog tweets dataset there are names who do not look like dogs names

**Code**

In [261]:
not_dog_name_index = dogs_tweets_clean[dogs_tweets_clean.name.str.islower()==True].index
dogs_tweets_clean = dogs_tweets_clean.drop(not_dog_name_index)

0

**Test**

In [262]:
dogs_tweets_clean[dogs_tweets_clean.name.str.islower()==True].name.count()

0

**Define**

In the dog tweets dataset there are few post with ratings which are not really dogs ratings

* *835152434251116546*

* *746906459439529985*

* *835246439529840640*

**Code**

In [270]:
jokes_not_dogs_arr = dogs_tweets_clean.query(
    "tweet_id in ('835246439529840640', '746906459439529985', '835152434251116546')").index
dogs_tweets_clean = dogs_tweets_clean.drop(jokes_not_dogs_arr)

**Test**

In [274]:
dogs_tweets_clean.query("tweet_id in ('835246439529840640', '746906459439529985', '835152434251116546')").tweet_id.count()

0

**Define**

For better reading the following column names need to be change:

1.Dog tweets: Extended url name need to be change to tweets_url

2.Predictions: jpg_url change for clear name, like prediction_img_url 

**Code**

In [276]:
dogs_tweets_clean.rename(columns={'expanded_urls':'tweets_url'}, inplace=True)
predictions_clean.rename(columns={'jpg_url':'prediction_img_url'}, inplace=True)

**Test**

In [277]:
print(dogs_tweets_clean.info())
print(predictions_clean.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2527 entries, 0 to 8151
Data columns (total 14 columns):
 #   Column                      Non-Null Count  Dtype              
---  ------                      --------------  -----              
 0   tweet_id                    2527 non-null   object             
 1   in_reply_to_status_id       81 non-null     float64            
 2   in_reply_to_user_id         81 non-null     float64            
 3   timestamp                   2527 non-null   datetime64[ns, UTC]
 4   source                      2527 non-null   object             
 5   text                        2527 non-null   object             
 6   retweeted_status_id         202 non-null    float64            
 7   retweeted_status_user_id    202 non-null    float64            
 8   retweeted_status_timestamp  202 non-null    object             
 9   tweets_url                  2466 non-null   object             
 10  rating_numerator            2527 non-null   int64           

**Define**

Dog tweets: remove retweets as they can be the same dog. There are 202 retweets.

**Code**

In [286]:
dogs_tweets_clean = dogs_tweets_clean[dogs_tweets_clean.retweeted_status_id.isna()]

**Test**

In [289]:
dogs_tweets_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2325 entries, 0 to 8151
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype              
---  ------                    --------------  -----              
 0   tweet_id                  2325 non-null   object             
 1   in_reply_to_status_id     81 non-null     float64            
 2   in_reply_to_user_id       81 non-null     float64            
 3   timestamp                 2325 non-null   datetime64[ns, UTC]
 4   source                    2325 non-null   object             
 5   text                      2325 non-null   object             
 6   retweeted_status_user_id  0 non-null      float64            
 7   tweets_url                2265 non-null   object             
 8   rating_numerator          2325 non-null   int64              
 9   rating_denominator        2325 non-null   int64              
 10  name                      1537 non-null   object             
 11  dog_type         

In [288]:
col_to_drop = ['retweeted_status_timestamp', 'retweeted_status_id', 'retweeted_status_timestamp', 'retweeted_status_user_id']
dogs_tweets_clean.drop(columns=col_to_drop, inplace=True)

C:\Users\user\anaconda3\envs\Udacity\lib\site-packages\pandas\core\frame.py:4160: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


**Define**

Predictions: Remove duplicate urls

**Code**

In [292]:
predictions_clean.drop_duplicates(subset=['prediction_img_url'], inplace=True)

**Test**

In [293]:
predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2009 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tweet_id            2009 non-null   object 
 1   prediction_img_url  2009 non-null   object 
 2   img_num             2009 non-null   int64  
 3   p1                  2009 non-null   object 
 4   p1_conf             2009 non-null   float64
 5   p1_dog              2009 non-null   bool   
 6   p2                  2009 non-null   object 
 7   p2_conf             2009 non-null   float64
 8   p2_dog              2009 non-null   bool   
 9   p3                  2009 non-null   object 
 10  p3_conf             2009 non-null   float64
 11  p3_dog              2009 non-null   bool   
dtypes: bool(3), float64(3), int64(1), object(5)
memory usage: 162.8+ KB


In [295]:
predictions_clean['prediction_img_url'].duplicated().sum()

0

**Define**

In dog tweets there are more then 1 rows with incorrect values, where int need to be float

**Code**

In [297]:
# first lets convert to float from int

dogs_tweets_clean['rating_numerator'] = dogs_tweets_clean['rating_numerator'].astype(float)
dogs_tweets_clean['rating_denominator'] = dogs_tweets_clean['rating_denominator'].astype(float)

<ipython-input-297-3ba80e920bfd>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dogs_tweets_clean['rating_numerator'] = dogs_tweets_clean['rating_numerator'].astype(float)
<ipython-input-297-3ba80e920bfd>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dogs_tweets_clean['rating_denominator'] = dogs_tweets_clean['rating_denominator'].astype(float)


In [300]:
# check for float ratings

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets_clean[dogs_tweets_clean['text'].str.contains(r"(\d+\.\d*\/\d+)")][
        ['text', 'rating_numerator', 'rating_denominator']])

C:\Users\user\anaconda3\envs\Udacity\lib\site-packages\pandas\core\strings.py:2001: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


text  \
45                             This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948   
695                                    This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
1689                                                         I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace   
1712                                                         Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD   
5475  This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   

      rating_numerator  rating_denominator  
45                 5.0                10.0  
695               75.0                10.0  
763               27.0                10.0  
1689               5.0                10.0  
1712              26.0                10.0  
5475              27.0                10.0

In [301]:
dogs_tweets_clean.loc[45, 'rating_numerator'] = 13.5
dogs_tweets_clean.loc[695, 'rating_numerator'] = 9.75
dogs_tweets_clean.loc[763, 'rating_numerator'] = 11.27
dogs_tweets_clean.loc[1689, 'rating_numerator'] = 9.5
dogs_tweets_clean.loc[1712, 'rating_numerator'] = 11.26
dogs_tweets_clean.loc[5475, 'rating_numerator'] = 11.26

C:\Users\user\anaconda3\envs\Udacity\lib\site-packages\pandas\core\indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


**Test**

In [302]:
# need to check one by one

with pd.option_context('display.max_rows', None, 'display.max_columns', None, 'max_colwidth', 200):
    display(dogs_tweets_clean[dogs_tweets_clean['text'].str.contains(r"(\d+\.\d*\/\d+)")][
        ['text', 'rating_numerator', 'rating_denominator']])

C:\Users\user\anaconda3\envs\Udacity\lib\site-packages\pandas\core\strings.py:2001: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  return func(self, *args, **kwargs)


text  \
45                             This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948   
695                                    This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   
763   This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
1689                                                         I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace   
1712                                                         Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD   
5475  This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   

      rating_numerator  rating_denominator  
45               13.50                10.0  
695               9.75                10.0  
763              11.27                10.0  
1689              9.50                10.0  
1712             11.26                10.0  
5475             11.26                10.0

# Create New Dataset
<a name="storing"></a>

In [303]:
tweets_prediction_merge = dogs_tweets_clean.merge(predictions_clean, left_on='tweet_id', right_on='tweet_id')
twitter_archive_master = tweets_prediction_merge.merge(tweet_info_clean, left_on='tweet_id', right_on='tweet_id')
twitter_archive_master.shape

(2127, 25)

In [304]:
twitter_archive_master.head(1)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   

                  timestamp              source  \
0 2017-08-01 16:23:56+00:00  Twitter for iPhone   

                                                text  \
0  This is Phineas. He's a mystical boy. Only eve...   

   retweeted_status_user_id  \
0                       NaN   

                                          tweets_url  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...              13.0   

   rating_denominator  ...   p1_conf p1_dog     p2   p2_conf p2_dog      p3  \
0                10.0  ...  0.097049  False  bagel  0.085851  False  banana   

   p3_conf p3_dog  retweet_count  favorite_count  
0  0.07611  False           8853           39467  

[1 rows x 25 columns]

In [305]:
twitter_archive_master.to_csv('twitter_archive_master.csv', index=False)

# Insights 
<a name="insights"></a>

## Insight 1
<a name="insight1"></a>

## Insight 2
<a name="insight2"></a>

## Insight 3 
<a name="insight3"></a>